Handle SIGTERM asynchronously

davidkohn88 opened the pull request:
SIGTERM/SIGINT shouldn't cause an exit in the handler functions as you might interrupt something important (like having a spinlock etc)
Instead, we should only set vars in the handler functions, then handle them in a separate function built for that purpose.

JLockerman commented on src/loader/bgw_launcher.c:
```diff
@@ -588,25 +588,55 @@ launcher_handle_message(HTAB *db_htab)
+	GOT_SIGTERM,
+}			ShutdownSignal;
+
+static volatile sig_atomic_t shutdown_signal = NO_SIGNAL;
```
For when coverity complains about these line: yes, this is probably the safest way to set a signal-variable; if we really don't want to use an `enum` we can just `#define` the constants, for the numbers we're using it should amount to the same thing

erimatnor commented on src/loader/bgw_launcher.c:
```diff
@@ -588,25 +588,55 @@ launcher_handle_message(HTAB *db_htab)
+			if (CritSectionCount != 0)
+				ereport(PANIC, (errmsg("TimescaleDB background worker launcher terminated while in a critical section")));
+			ereport(LOG, (errmsg("TimescaleDB background worker launcher terminated by administrator command. Launcher will not restart after exiting")));
+			proc_exit(0);
```
Why do we explicitly exit here as opposed to just returning from the main function? I think breaking out of the main loop is what you would normally do.  IMO opinion, the code is harder to understand if we exit at arbitrary places rather than just executing the main function to its end, especially in case you do things like cleanup, etc.

erimatnor commented on src/loader/bgw_launcher.c:
```diff
@@ -646,7 +676,7 @@ ts_bgw_cluster_launcher_main(PG_FUNCTION_ARGS)
 	CHECK_FOR_INTERRUPTS();
 
-	while (true)
+	while (shutdown_signal == NO_SIGNAL)
```
Can't we call `CHECK_FOR_INTERRUPTS` the first thing we do in the loop and then you can just replace the two calls (above and at the end of the loop)  with one.

erimatnor commented on src/loader/bgw_launcher.c:
```diff
@@ -588,25 +588,55 @@ launcher_handle_message(HTAB *db_htab)
+			 * The launcher should never be in a critical section, but if it
+			 * is we have no choice but to PANIC since it is unsafe to exit.
+			 */
+			if (CritSectionCount != 0)
```
Do we need this code? I don't see how it can happen since this function is only called from the end of the main function, which is not in a critical section?

cevian replied:
Agreed with Erik. I would just Assert(CritSectionCount != 0)

davidkohn88 commented:
Decided to change the approach and simply use the Postgres `die()` functions with a small wrapper around them, similar to how @cevian solved in #734. I think that also addressed your nits.

JLockerman commented on src/loader/bgw_launcher.c:
```diff
@@ -35,6 +35,8 @@
 #include "bgw_message_queue.h"
 #include "bgw_launcher.h"
 
+#include <tcop/tcopprot.h>
```
I don't think this is needed anymore, it was just there for debugging

davidkohn88 replied:
This is needed for die(), added clarification and put it in a better place in the headers

JLockerman commented on src/loader/bgw_launcher.c:
```diff
@@ -695,20 +708,26 @@ database_is_template_check(void)
  */
 extern Datum
 ts_bgw_db_scheduler_entrypoint(PG_FUNCTION_ARGS)
+
```
extra line

JLockerman commented on src/loader/bgw_launcher.c:
```diff
@@ -588,23 +590,26 @@ launcher_handle_message(HTAB *db_htab)
-bgw_sigint(SIGNAL_ARGS)
-{
-	ereport(ERROR, (errmsg("TimescaleDB background worker launcher canceled by administrator command. Launcher will restart after exiting")));
+	ereport(LOG,
```
This might cause issues in the bgw tests when we send background worker output to a database in a transaction; is there a chance we'll reenter the signal handler?

erimatnor commented on src/loader/bgw_launcher.c:
```diff
@@ -588,23 +591,26 @@ launcher_handle_message(HTAB *db_htab)
+	ereport(LOG,
+			(errcode(ERRCODE_ADMIN_SHUTDOWN),
+			 errmsg("terminating TimescaleDB background worker launcher due to administrator command")));
+	die(postgres_signal_arg);
```
In my understanding, `die` doesn't call `pg_exit()` so I don't get where we exit now since I see no exit condition in the main loop.

erimatnor commented on src/loader/bgw_launcher.c:
```diff
@@ -644,12 +650,11 @@ ts_bgw_cluster_launcher_main(PG_FUNCTION_ARGS)
 	{
 		int			wl_rc;
 
+		CHECK_FOR_INTERRUPTS();
```
I guess I am missing something, but don't we need to check for an exit condition that is set by `die()` here?

I see in `die()` that it exits immediately if in single user mode, which might be the case here. But that seems very implicit so shouldn't we check `ProcDiePending` here at least for clarity?

davidkohn88 commented:
This should fix any other remaining issues in #733 and #707 that weren't addressed in PRs #723  and #715 .
